src/cpp/alohalytics.cc

```diff
@@ -161,86 +174,138 @@ static inline void LogEventImpl(AlohalyticsBaseEvent const & event, THundredKilo
   messages_queue.PushMessage(sstream.str());
 }
 
-void Stats::LogEvent(std::string const & event_name) {
+void Stats::LogEvent(std::string const & event_name, uint32_t channels_mask) {
   LOG_IF_DEBUG("LogEvent:", event_name);
   if (enabled_) {
     AlohalyticsKeyEvent event;
     event.key = event_name;
-    LogEventImpl(event, messages_queue_);
+    for (uint32_t i = 0; i < channels_.size(); ++i) {
+      if (channels_mask & ChannelMask(i))
+        LogEventImpl(event, channels_[i]->messages_queue_);
+    }
   }
 }
 
-void Stats::LogEvent(std::string const & event_name, Location const & location) {
+void Stats::LogEvent(std::string const & event_name, Location const & location, uint32_t channels_mask) {
   LOG_IF_DEBUG("LogEvent:", event_name, location.ToDebugString());
   if (enabled_) {
     AlohalyticsKeyLocationEvent event;
     event.key = event_name;
     event.location = location;
-    LogEventImpl(event, messages_queue_);
+    for (uint32_t i = 0; i < channels_.size(); ++i) {
+      if (channels_mask & ChannelMask(i))
+        LogEventImpl(event, channels_[i]->messages_queue_);
+    }
   }
 }
 
-void Stats::LogEvent(std::string const & event_name, std::string const & event_value) {
+void Stats::LogEvent(std::string const & event_name, std::string const & event_value, uint32_t channels_mask) {
   LOG_IF_DEBUG("LogEvent:", event_name, "=", event_value);
   if (enabled_) {
     AlohalyticsKeyValueEvent event;
     event.key = event_name;
     event.value = event_value;
-    LogEventImpl(event, messages_queue_);
+    for (uint32_t i = 0; i < channels_.size(); ++i) {
+      if (channels_mask & ChannelMask(i))
+        LogEventImpl(event, channels_[i]->messages_queue_);
+    }
   }
 }
 
-void Stats::LogEvent(std::string const & event_name, std::string const & event_value, Location const & location) {
+void Stats::LogEvent(std::string const & event_name, std::string const & event_value, Location const & location,
+                     uint32_t channels_mask) {
   LOG_IF_DEBUG("LogEvent:", event_name, "=", event_value, location.ToDebugString());
   if (enabled_) {
     AlohalyticsKeyValueLocationEvent event;
     event.key = event_name;
     event.value = event_value;
     event.location = location;
-    LogEventImpl(event, messages_queue_);
+    for (uint32_t i = 0; i < channels_.size(); ++i) {
+      if (channels_mask & ChannelMask(i))
+        LogEventImpl(event, channels_[i]->messages_queue_);
+    }
```

Comment:
Can we factor this out to LogEventImpl(event, channels_mask)?

Reply:
done